doc/memoize-clean.1.md
---
title: memoize-clean
section: 1
header: User Manual
footer: memoize-clean of Memoize v1.4.1
date: December 02, 2024
hyphenate: false
---

# NAME
memoize-clean.pl, memoize-clean.py - Remove (stale) memo and extern files


# SYNOPSIS
**memoize-clean.pl** [*OPTIONS*] [*document1.mmz* ...]

**memoize-clean.py** [*OPTIONS*] [*document1.mmz* ...]


# DESCRIPTION

**memoize-clean** is a script accompanying Memoize, a TeX package which allows
the author to reuse the results of compilation-intensive code such as TikZ
pictures.

By default, this script removes stale memo and extern files.  A stale memo or
extern is an extern produced during a compilation of a previous version of the
document which was neither used nor produced during the last compilation of the
document.  Typically, stale files arise when we change the memoized code (or
its context).

**memoize-clean.pl** removes all memo and extern files with prefixes mentioned
in the listed **.mmz** files and by options **\--prefix** which are not
explicitly mentioned in the **.mmz** files.

Before deleting anything, **memoize-clean.pl** lists all the files it would
delete and asks for confirmation.

# OPTIONS

**-p, \--prefix**
: Add a memo/extern prefix for cleaning. This option may be given multiple
    times.

**-a, \--all**
: Remove all memos and externs, rather than only the stale ones.

**-y, \--yes**
: Do not ask for confirmation.

**-q, \--quiet**
: Don't describe what's happening.

**-h, \--help**
: Show help.

**-V, \--version**
: Show the Memoize version number and exit.

# SEE ALSO

[Memoize manual](https://ctan.org/pkg/memoize), section 6.6.3.


CHANGELOG.md
# Memoize changelog

For the development history, see [Memoize's GitHub
repository](https://github.com/sasozivanovic/memoize).

## 2024/12/02 v1.4.1

* Bugfix: the landscape support introduced in 1.4.0 broke Memoize under LuaTeX.

## 2024/11/24 v1.4.0

* New features:
  * Support Beamer modes.
  * Support externs typeset on landscape pages.
  * Implement `salt` and related keys (used by Beamer mode support).

* Minor changes:
  * Allow changing `prefix` in `\mmznext`.
  * To better support user interaction with foreign package support, any
    unknown `/mmz` keys issued in the preamble are retried at `begindocument/end`.

* Bugfixes:
  * Allow for (expandable) macros in reference keys advised by `auto/ref`.
  * Fix a regular expression in `memoize-clean.py`.

* Documentation:
  * Improve the troubleshooting sections (2.9, 3.1).
  * Update and rewrite installation instructions.
  * Fix some typos and formatting issues.

* Under the hood:
  * The Beamer support code now resides in a separate file, `memoize-beamer.code.tex`.
  * Make PGF support produce nicer cc-memos.
	
* Build system:
  * A new, makefile-controlled system for compiling examples, which are now
    compiled prior to compiling the manual.
  * Implement a different way of automatically changing the version number in
    the documentation.
  * Compile the documentation through dedicated `latexmkrc` config files, using
    an adapted `memoize_latexmkrc` example config for the manual.
  * Partially change installation-related parts of the Makefile.
  * Include the missing makefiles in CTAN & TDS zip files.

## 2024/04/02 v1.3.0

* New defaults:
  * `memo dir` is now in effect by default.
  * `mkdir command` is now initialized to `memoize-extract.pl --mkdir` even
    when `extract=no` or `extract=tex`.
* Update the manual to reflect the new defaults (plus many minor improvements).
* Bugfixes:
  * The extraction scripts (affecting Windows users): properly escape `\` in
	log messages.
  * Biblatex support: `\volcites` now works as advertised.

## 2024/03/15 v1.2.0

* Biblatex support:
  * Allow for entries containing verbatim material.
  * Support `\volcite` commands.
  * Implement `biblatex ccmemo cite`.
  * Submit all known citation commands to `auto`-keys `(vol)cite(s)`.
	* The support must be explicitly loaded by `\mmzset{biblatex}`.
* Minor changes:
	* Separate generic PGF support out of TikZ support.
	* Support `latexmk`.
	* Drop the obsolete workaround for package `morewrites`.
	* Clear Memoize's `begindocument` hooks after executing them.
* Documentation:
  * Introduce section "Support for specific classes and packages".
  * Improve the documentation of argument specification accepted by CollArgs'
		command `\CollectArguments` and Advice's key `args`, in particular with
		reference to the fact that since 2020, the functionality of package
		`xparse` is mostly integrated into the LaTeX kernel.
	* Add a note about `TEXLIVE_WINDOWS_TRY_EXTERNAL_PERL`.
	* Various minor changes.
	
## 2024/01/21 v1.1.2

* Fix a bug in Biblatex support.

## 2024/01/16 v1.1.1

* Fix a bug where, under `no memo dir`, Memoize was checking whether the extern
  exists in the root folder.

## 2024/01/02 v1.1.0

* Improve the extraction scripts:
  * respect `TEXMF_OUTPUT_DIRECTORY`;
  * respect `openin_any` and `openout_any`;
  * implement `--format`;
  * improve error reporting;
  * drop the `Path::Class` dependency for the Perl script;
  * allow for `PDF::Builder` in the Perl script;
  * implement `--library` in the Perl script;
  * set an appropriate exit code on exit;
  * and several further minor changes.

* Remove key `path` in favour of `prefix`. 

* `mkdir` is now initially `true`, but the directory is only created if `mkdir
  command` is non-empty (and it is empty initially).  The definition of `(no)
  memo dir` is accordingly simpler.

* The directory name is now appended to the value `mkdir command` when
  constructing the system call.

* A workaround for compatibility with package `morewrites`.

* Process package options using the new LaTeX mechanism to avoids the issue of
  spaces in package options.  The remaining issue of `/` is addressed by
  implementing option `options`.

* Add the missing commands to `nomemoize` and `memoizable`, and implement a
  generic variant of the latter (`memoizable.code.tex`).

* Implement auto-key `to context`.

* Write a c-memo even upon abortion.

* Demote warning messages "memoization aborted" & "marked as unmemoizable" to
  info messages.

* Implement biblatex support.

* Support `\DiscardShipoutBox`.

* Advance the counter underlying `\pgfpictureid` when utilizing a tikzpicture
  (`memoize tikz`).

* Remove the `\pgfsys@getposition` hack for `tikzpicture`s.

## 2023/10/10 v1.0.0

* A complete, fully documented reimplementation.

## 2020/07/17 v0.1

* The proof of concept, available only at GitHub.


memoize-clean.py
#!/usr/bin/env python

# This file is a part of Memoize, a TeX package for externalization of
# graphics and memoization of compilation results in general, available at
# https://ctan.org/pkg/memoize and https://github.com/sasozivanovic/memoize.
#
# Copyright (c) 2020- Saso Zivanovic <saso.zivanovic@guest.arnes.si>
#
# This work may be distributed and/or modified under the conditions of the
# LaTeX Project Public License, either version 1.3c of this license or (at
# your option) any later version.  The latest version of this license is in
# https://www.latex-project.org/lppl.txt and version 1.3c or later is part of
# all distributions of LaTeX version 2008 or later.
#
# This work has the LPPL maintenance status `maintained'.
# The Current Maintainer of this work is Saso Zivanovic.
# 
# The files belonging to this work and covered by LPPL are listed in
# <texmf>/doc/generic/memoize/FILES.

__version__ = '2024/10/25 v1.4.0'

import argparse, re, sys, pathlib, os

parser = argparse.ArgumentParser(
    description="Remove (stale) memo and extern files.",
    epilog = "For details, see the man page or the Memoize documentation "
             "(https://ctan.org/pkg/memoize)."
)
parser.add_argument('--yes', '-y', action = 'store_true',
                    help = 'Do not ask for confirmation.')
parser.add_argument('--all', '-a', action = 'store_true',
                    help = 'Remove *all* memos and externs.')
parser.add_argument('--quiet', '-q', action = 'store_true')
parser.add_argument('--prefix', '-p', action = 'append', default = [],
    help = 'A path prefix to clean; this option can be specified multiple times.')
parser.add_argument('mmz', nargs= '*', help='.mmz record files')
parser.add_argument('--version', '-V', action = 'version',
                    version = f"%(prog)s of Memoize " + __version__)
args = parser.parse_args()

re_prefix = re.compile(r'\\mmzPrefix *{(.*?)}')
re_memo = re.compile(r'%? *\\mmz(?:New|Used)(?:CC?Memo|Extern) *{(.*?)}')
re_endinput = re.compile(r' *\\endinput *$')

prefixes = set(pathlib.Path(prefix).resolve() for prefix in args.prefix)
keep = set()

# We loop through the given .mmz files, adding prefixes to whatever manually
# specified by the user, and collecting the files to keep.
for mmz_fn in args.mmz:
    mmz = pathlib.Path(mmz_fn)
    mmz_parent = mmz.parent.resolve()
    try:
        with open(mmz) as mmz_fh:
            prefix = ''
            endinput = False
            empty = None
            for line in mmz_fh:
                line = line.strip()
                
                if not line:
                    pass
                
                elif endinput:
                    raise RuntimeError(
                        rf'Bailing out, '
                        rf'\endinput is not the last line of file {mmz_fn}.')
                
                elif m := re_prefix.match(line):
                    prefix = m[1]
                    prefixes.add( (mmz_parent/prefix).resolve() )
                    if empty is None:
                        empty = True

                elif m := re_memo.match(line):
                    if not prefix:
                        raise RuntimeError(
                            f'Bailing out, no prefix announced before file "{m[1]}".')
                    if not m[1].startswith(prefix):
                        raise RuntimeError(
                            f'Bailing out, prefix of file "{m[1]}" does not match '
                            f'the last announced prefix ({prefix}).')
                    keep.add((mmz_parent / m[1]))
                    empty = False

                elif re_endinput.match(line):
                    endinput = True
                    continue

                else:
                    raise RuntimeError(fr"Bailing out, "
                        fr"file {mmz_fn} contains an unrecognized line: {line}")
            
        if empty and not args.all:
            raise RuntimeError(fr'Bailing out, file {mmz_fn} is empty.')

        if not endinput and empty is not None and not args.all:
            raise RuntimeError(
                fr'Bailing out, file {mmz_fn} does not end with \endinput; '
                fr'this could mean that the compilation did not finish properly. '
                fr'You can only clean with --all.'
            )
        
    # It is not an error if the file doesn't exist.
    # Otherwise, cleaning from scripts would be cumbersome.
    except FileNotFoundError:
        pass

tbdeleted = []
def populate_tbdeleted(folder, basename_prefix):
    re_aux = re.compile(
        re.escape(basename_prefix) + 
        r'[0-9A-F]{32}(?:-[0-9A-F]{32})?'
        r'(?:-[0-9]+)?(?:\.memo|(?:-[0-9]+)?\.pdf|\.log)$')
    try:
        for f in folder.iterdir():
            if re_aux.match(f.name) and (args.all or f not in keep):
                tbdeleted.append(f)
    except FileNotFoundError:
        pass

for prefix in prefixes:
    # "prefix" is interpreted both as a directory (if it exists) and a basename prefix.
    if prefix.is_dir():
        populate_tbdeleted(prefix, '')
    populate_tbdeleted(prefix.parent, prefix.name)

allowed_dirs = [pathlib.Path().absolute()] # todo: output directory
deletion_not_allowed = [f for f in tbdeleted if not f.is_relative_to(*allowed_dirs)]
if deletion_not_allowed:
    raise RuntimeError("Bailing out, "
        "I was asked to delete these files outside the current directory:\n" +
        "\n".join(str(f) for f in deletion_not_allowed))
        
_cwd_absolute = pathlib.Path().absolute()
def relativize(path):
    try:
        return path.relative_to(_cwd_absolute)
    except ValueError:
        return path

if tbdeleted:
    tbdeleted.sort()
    if not args.yes:
        print('I will delete the following files:')
        for f in tbdeleted:
            print(relativize(f))
        print("Proceed (y/n)? ")
        a = input()
    if args.yes or a == 'y' or a == 'yes':
        for f in tbdeleted:
            if not args.quiet:
                print("Deleting", relativize(f))
            try:
                f.unlink()
            except FileNotFoundError:
                print(f"Cannot delete {f}")
    else:
        print("Bailing out.")
elif not args.quiet:
    print('Nothing to do, the directory seems clean.')

# Local Variables:
# fill-column: 79
# after-save-hook: py2dtx
# End:
